pkg/ebpf/dns_cache.go

```diff
@@ -0,0 +1,204 @@
+package ebpf
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type reverseDNSCache struct {
+	mux  sync.Mutex
+	data map[util.Address]*dnsCacheVal
+	exit chan struct{}
+	ttl  time.Duration
+	size int
+
+	// Telemetry
+	len      int64
+	lookups  int64
+	resolved int64
+}
+
+func newReverseDNSCache(size int, ttl, expirationPeriod time.Duration) *reverseDNSCache {
+	cache := &reverseDNSCache{
+		data: make(map[util.Address]*dnsCacheVal),
+		exit: make(chan struct{}),
+		ttl:  ttl,
+		size: size,
+	}
+
+	ticker := time.NewTicker(expirationPeriod)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				cache.expire()
+			case <-cache.exit:
+				return
+			}
+		}
+	}()
+	return cache
+}
+
+func (c *reverseDNSCache) Add(translation *translation, now time.Time) bool {
+	if translation == nil {
+		return false
+	}
+
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	if len(c.data) >= c.size {
+		return false
+	}
+
+	exp := now.Add(c.ttl).Unix()
+	for addr := range translation.ips {
+		val, ok := c.data[addr]
+		if ok {
+			val.expiration = exp
+			val.merge(translation.name)
+			continue
+		}
+
+		c.data[addr] = &dnsCacheVal{names: []string{translation.name}, expiration: exp}
+	}
+
+	// Update cache length for telemetry purposes
+	atomic.StoreInt64(&c.len, int64(len(c.data)))
+
+	return true
+}
+
+func (c *reverseDNSCache) Get(conns []ConnectionStats, now time.Time) map[util.Address][]string {
+	if len(conns) == 0 {
+		return nil
+	}
+
+	var (
+		ipToNames  = make(map[util.Address][]string)
+		visited    = make(map[util.Address]struct{})
+		expiration = now.Add(c.ttl).Unix()
+	)
+
+	collectNamesForIP := func(addr util.Address) {
+		if _, ok := visited[addr]; ok {
+			return
+		}
+		visited[addr] = struct{}{}
+
+		if names := c.getNamesForIP(addr, expiration); names != nil {
+			ipToNames[addr] = names
+		}
+	}
+
+	c.mux.Lock()
+	for _, conn := range conns {
+		collectNamesForIP(conn.Source)
```

Comment:
Another trick that would likely help a lot here (especially for local connections) is to compare Source and Dest.

```
collectNamesForIP(conn.Source)
if conn.Source != conn.Dest {
  collectNamesForIP(conn.Dest)
}
```

Reply:
Once you get rid of `visted`, you can probably just inline 
```
	if names := c.getNamesForIP(addr, expiration); names != nil {
			ipToNames[addr] = names
		}
```
twice here instead of the local function.

Reply:
Considering what I said above about `visited`, I think this should stay as it is.